### PR TITLE
Host-assisted cloning runs after PVC is already cloned #534

### LINKED ISSUE #534: Host-assisted cloning runs after PVC is already cloned
/kind bug

**What happened**:
Host-assisted cloning runs on a PVC which was already cloned by the gluster subvol provisioner.

**What you expected to happen**:
Host-assisted cloning should not run on PVCs which already carry the CloneOf annotation.

**How to reproduce it (as minimally and precisely as possible)**:

- Begin with a development environment using ```make cluster-up cluster-sync```
- Create 10 PVCs and 10 Pods that use the PVCs in order to use up all PVs
-  Create PVC with CloneRequest annotation to clone one of the dummy PVCs
- The PVC is in pending state and clone pods are scheduled to run
- Simulate the completion of a clone by adding a CloneOf annotation to the PVC
- Delete one of the dummy PVCs
- Clone PVC binds to a PV and cloning pods run against the bound PVC

**Anything else we need to know?**:
The above steps simulate the behavior without needing to deploy the gluster subvol provisioner.  I think the proper fix for this issue is to not schedule the clone pods unless the target PVC is either Bound or Pending/WaitFirstConsumer.  At that point you can check for the CloneOf annotation and if it's missing you will be confident that the provisioner will not actually do a clone itself.

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`):
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by zvikorn**
Signed-off-by: tavni <tavni@redhat.com>

**What this PR does / why we need it**:
This PR fixes issue #534 
Host-assisted cloning runs on a PVC which was already cloned by the gluster subvol provisioner.

**Which issue(s) this PR fixes** :
Fixes #534 

**Special notes for your reviewer**:
To test the fix works (we test it with local storage since gluster subvol is not necessary for this test):
- Begin with a development environment using make cluster-up cluster-sync
- Create and deploy a storageclase with no volumeBindingMode field since the existing PV's contain 
  this field with the value 'WaitForFirstConsumer'.
- Create a PV with the storageclass name you just created. Do NOT deploy the PV yet.
- Create and deploy 10 PVCs to import disk image. 10 importer Pods will run and complete. All 10 
  PV's are bound to the 10 PVC's.
- Create and deploy a cloner PVC with CloneRequest annotation and storageclass with the name of 
   the storageclass you created above to clone one of the dummy PVCs. 
- Since there is no available PV, the PVC wil be in 'Pending' phase. Cloner pods will NOT be spawned 
  with this fix. This is the important thing to verify! Cloner pods are NOT spawned.
- Edit the deployed cloner PVC and add the "k8s.io/CloneOf: true" annotation to indicate that the clone 
  was done by external provisioner (smart cloning).
- Deploy the PV you created above to make the cloner PVC bound to this PV.
- VERIFY cloner pods are NOT spawned (this is because the CloneOf annotation).

```release-note
Fixing issue #534 
```

### CODE REVIEW

File: pkg/controller/clone-controller.go [hunk 1]
```diff
@@ -230,6 +231,23 @@ func (cc *CloneController) syncPvc(key string) error {
 	if !checkPVC(pvc, AnnCloneRequest) {
 		return nil
 	}
+
+	pvcPhase := pvc.Status.Phase
+	glog.V(3).Infof("PVC phase for PVC \"%s/%s\" is ", pvc.Namespace, pvc.Name, pvcPhase)
```
Comment (awels) [inline pkg/controller/clone-controller.go]: I think you are missing a %s in the log message.

File: pkg/controller/clone-controller.go [hunk 2]
```diff
@@ -230,6 +231,23 @@ func (cc *CloneController) syncPvc(key string) error {
...
+	storageClassName := *pvc.Spec.StorageClassName
+	glog.V(3).Infof("storageClassName used by PVC \"%s/%s\" is \"%s\"", pvc.Namespace, pvc.Name, storageClassName)
+	storageclass, err := cc.clientset.StorageV1().StorageClasses().Get(storageClassName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	//Do not schedule the clone pods unless the target PVC is either Bound or Pending/WaitFirstConsumer.
+	if !(pvcPhase == v1.ClaimBound || (pvcPhase == v1.ClaimPending && *storageclass.VolumeBindingMode == storageV1.VolumeBindingWaitForFirstConsumer)) {
+		glog.V(3).Infof("PVC \"%s/%s\" is either not bound or is in pending phase and VolumeBindingModein is not VolumeBindingWaitForFirstConsumer."+
```
Comment (awels) [inline pkg/controller/clone-controller.go]: ... and VolumeBindingMode is ...

File: pkg/controller/clone-controller.go (hunk 1 above)
Comment (zvikorn) [reply pkg/controller/clone-controller.go]: Thanks.

File: pkg/controller/clone-controller.go (hunk 2 above)
Comment (zvikorn) [reply pkg/controller/clone-controller.go]: added log

File: pkg/controller/clone-controller.go [hunk 3]
```diff
@@ -230,6 +231,23 @@ func (cc *CloneController) syncPvc(key string) error {
 	if !checkPVC(pvc, AnnCloneRequest) {
 		return nil
 	}
+
+	pvcPhase := pvc.Status.Phase
+	glog.V(3).Infof("PVC phase for PVC \"%s/%s\" is ", pvc.Namespace, pvc.Name, pvcPhase)
+	storageClassName := *pvc.Spec.StorageClassName
```
[the remark this answers is not in the document]
Comment (zvikorn) [reply pkg/controller/clone-controller.go]: Fixed. Ready for another review. Thank you!

**Pull request merged**